log4j-api/src/main/java/org/apache/logging/log4j/CloseableThreadContext.java

```diff
@@ -128,6 +153,37 @@ public Instance put(final String key, final String value) {
         }
 
         /**
+         * Populates the Thread Context Map with the supplied key/value pairs. Any existing keys in the
+         * {@link ThreadContext} will be replaced with the supplied values, and restored back to their original value when
+         * the instance is closed.
+         *
+         * @param values The map of key/value pairs to be added
+         * @return a new instance that will back out the changes when closed.
+         * @since 2.7.1
+         */
+        public Instance putAll(final Map<String, String> values) {
```

Comment:
The atomicity does not cover map and stack together, only the context map. So to correct myself: `ThreadContext.putAll()` will do an atomic update of the context map.

It is also about performance since the context map is often implemented in a copy-on-write fashion.